// @cmsgov/design-system/dist/preact-components/cjs/CloseButton/CloseButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.CloseButton = void 0;
var _classnames = _interopRequireDefault(require("classnames"));
var _useId = _interopRequireDefault(require("../utilities/useId"));
var _Icons = require("../Icons");
var _jsxRuntime = require("preact/compat/jsx-runtime");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
/**
 *
 */
const CloseButton = _ref => {
  let {
    className,
    id: idProp,
    ariaHidden = false,
    ...buttonAttributes
  } = _ref;
  const id = (0, _useId.default)('close-button--', idProp);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
    type: "button",
    ...buttonAttributes,
    className: (0, _classnames.default)('ds-c-close-button', className),
    id: id,
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Icons.CloseIconThin, {
      ariaHidden: ariaHidden,
      id: `${id}__icon`
    })
  });
};
exports.CloseButton = CloseButton;
var _default = CloseButton;
exports.default = _default;
