// @cmsgov/design-system/dist/react-components/cjs/TextField/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  TextInput: true,
  maskValue: true,
  unmaskValue: true,
  DATE_MASK: true,
  SSN_MASK: true,
  ZIP_MASK: true,
  PHONE_MASK: true,
  CURRENCY_MASK: true
};
Object.defineProperty(exports, "CURRENCY_MASK", {
  enumerable: true,
  get: function () {
    return _useLabelMask.CURRENCY_MASK;
  }
});
Object.defineProperty(exports, "DATE_MASK", {
  enumerable: true,
  get: function () {
    return _useLabelMask.DATE_MASK;
  }
});
Object.defineProperty(exports, "PHONE_MASK", {
  enumerable: true,
  get: function () {
    return _useLabelMask.PHONE_MASK;
  }
});
Object.defineProperty(exports, "SSN_MASK", {
  enumerable: true,
  get: function () {
    return _useLabelMask.SSN_MASK;
  }
});
Object.defineProperty(exports, "TextInput", {
  enumerable: true,
  get: function () {
    return _TextInput.default;
  }
});
Object.defineProperty(exports, "ZIP_MASK", {
  enumerable: true,
  get: function () {
    return _useLabelMask.ZIP_MASK;
  }
});
Object.defineProperty(exports, "maskValue", {
  enumerable: true,
  get: function () {
    return _maskHelpers.maskValue;
  }
});
Object.defineProperty(exports, "unmaskValue", {
  enumerable: true,
  get: function () {
    return _maskHelpers.unmaskValue;
  }
});
var _TextField = require("./TextField");
Object.keys(_TextField).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _TextField[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _TextField[key];
    }
  });
});
var _TextInput = _interopRequireDefault(require("./TextInput"));
var _maskHelpers = require("./maskHelpers");
var _useLabelMask = require("./useLabelMask");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
