// @cmsgov/design-system/dist/react-components/cjs/DateField/CustomDayPickerDayContent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CustomDayPickerDayContent = CustomDayPickerDayContent;
var _reactDayPicker = require("react-day-picker");
var _jsxRuntime = require("react/jsx-runtime");
/** Represent the props for the {@link DayContent} component. */

/**
 * Render the content of the day cell.
 */
/** Render the content of the day cell. */
function CustomDayPickerDayContent(props) {
  const {
    locale,
    classNames,
    styles,
    labels: {
      labelDay
    },
    formatters: {
      formatDay
    }
  } = (0, _reactDayPicker.useDayPicker)();
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
      "aria-hidden": "true",
      children: formatDay(props.date, {
        locale
      })
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
      className: classNames.vhidden,
      style: styles.vhidden,
      children: labelDay(props.date, props.activeModifiers, {
        locale
      })
    })]
  });
}
